package/repos/codeberg.py

```diff
@@ -0,0 +1,190 @@
+from dataclasses import dataclass
+import logging
+from typing import Optional
+
+from django.conf import settings
+import httpx
+
+from .base_handler import BaseHandler
+
+
+logger = logging.getLogger(__name__)
+
+
+REPOSITORY_URL = "https://codeberg.org/api/v1/repos/{repository}"
+COMMITS_URL = "https://codeberg.org/api/v1/repos/{repository}/commits"
+COLLABORATORS_URL = "https://codeberg.org/api/v1/repos/{repository}/collaborators"
+
+
+@dataclass
+class ForgejoMetadata:
+    archived: bool
+    archived_at: str
+    description: str
+    forks_count: int
+    watchers_count: int
+
+
+@dataclass
+class ForgejoCommit:
+    sha: str
+    created: str
+
+
+class CodebergHandler(BaseHandler):
+    title = "Codeberg"
+    url_regex = "(https|git)://codeberg.org/"
+    url = "https://codeberg.org/"
+    repo_regex = r"(?:https|git)://codeberg.org/[^/]*/([^/]*)"
+    slug_regex = repo_regex
+
+    def __init__(self):
+        self.api_token = settings.CODEBERG_TOKEN
+
+    def fetch_commits(self, package: "package.models.Package"):
+        from package.models import Commit  # Added here to avoid circular imports
+
+        for commit in _fetch_commits(package.repo_name()):
+            try:
+                _, created = Commit.objects.get_or_create(
+                    package=package,
+                    commit_hash=commit.sha,
```

Comment:
I am using https://codeberg.org/allauth/django-allauth as test repository.

`GitHubHandler` only uses the package and the commit date to create commits. allauth has several commits at the same time with means the stop condition below triggers after 27 commits instead of doing an initial sync of all 3574 commits. Passing the commit hash fixes this problem.

Is there a reason this isn't needed for GitHub? (mili / nano second precession instead of seconds?)

Reply:
Great find. This code (outside of my lint update) is 9 to 11 years old. I suspect it was an oversight. 

I'll have to do some testing, but I think we could switch to `commit_hash` if you want to switch this new repo handler to it over `commit_date`.

Reply:
Nice catch! 💯